File: tests/cross-verification/fmix64/_gen/gen.rs
// Independent Rust oracle: compute MurmurHash3 fmix64 over the canonical inputs and
// emit rust-output.json. Recomputes the finaliser with wrapping u64 arithmetic (does
// not depend on the zeta crate) so it is a genuine independent oracle.
use std::fs;
use std::path::Path;
fn fmix64(x: u64) -> u64 {
    let mut h = x;
    h ^= h >> 33;
    h = h.wrapping_mul(0xff51afd7ed558ccd);
    h ^= h >> 33;
    h = h.wrapping_mul(0xc4ceb9fe1a85ec53);
    h ^= h >> 33;
    h
}
fn main() {
    let inputs: [(&str, u64); 10] = [
        ("x-0", 0),
        ("x-1", 1),
        ("x-2", 2),
        ("x-10", 10),
        ("x-255", 255),
        ("x-u64max", 18446744073709551615),
        ("x-golden", 11400714819323198485),
        ("x-2pow63", 9223372036854775808),
        ("x-12345678901234567890", 12345678901234567890),
        ("x-1e18", 1000000000000000000),
    ];
    let mut s = String::from("{\n");
    for (i, (id, x)) in inputs.iter().enumerate() {
        let comma = if i < inputs.len() - 1 { "," } else { "" };
        s.push_str(&format!("  \"{}\": \"{}\"{}\n", id, fmix64(*x), comma));
    }
    s.push_str("}\n");
    // gen.rs is compiled+run with cwd = the _gen dir; write to the parent.
    let target = Path::new("..").join("rust-output.json");
    fs::write(target, s).expect("write rust-output.json");
    println!("wrote rust-output.json");
}

File: tests/cross-verification/generator-registry-id/_gen/gen.ts
// TS oracle for GeneratorRegistry.idOf — recomputes the content-addressed ZetaId
// for each canonical `name@version` and emits ts-output.json.
//
// This is a faithful independent re-derivation of the F# `hash128`/`idOf`
// (src/Core/GeneratorRegistry.fs): a two-lane FNV-1a-ish 128-bit fold over the
// string `"{name}@{version}"`, all arithmetic wrapping uint64, printed as 32-hex
// (h1 then h2). It does NOT import any zeta port — it is a genuine independent
// oracle, so cross-language agreement is real evidence, not re-serialisation.
//
// The emitted oracle tags itself `_source: generated-from-ir`: the registry id IS
// the IR coordinate. `rng.splitmix64@1` is the generator behind the splitmix64
// oracle, so this byte-lock is what ties that oracle's provenance to a real,
// reproducible, content-addressed registry row rather than a free-floating literal.

import { writeFileSync } from "node:fs";
import { join, dirname } from "node:path";

const MASK = (1n << 64n) - 1n;
const u64 = (x: bigint): bigint => x & MASK;
const PRIME = 0x100000001b3n;
const PRIME2 = 0x1000000021bn;

/** Mirror of GeneratorRegistry.hash128: two FNV-1a-ish u64 lanes over the string (decorrelated). */
function hash128(s: string): string {
  let h1 = 0xcbf29ce484222325n;
  let h2 = 0x84222325cbf29ce4n;
  for (const ch of s) {
    const c = BigInt(ch.codePointAt(0) ?? 0);
    h1 = u64((h1 ^ u64(c)) * PRIME);
    const rotated = u64((c << 31n) | (c >> 33n));
    h2 = u64((h2 ^ rotated) * PRIME2);
  }
  return h1.toString(16).padStart(16, "0") + h2.toString(16).padStart(16, "0");
}

/** Mirror of GeneratorRegistry.idOf. */
function idOf(name: string, version: number): string {
  return hash128(`${name}@${version}`);
}

// (name, version) inputs — keyed by the canonical id `name@version`.
const inputs: { id: string; name: string; version: number }[] = [
  { id: "rng.splitmix64@1", name: "rng.splitmix64", version: 1 },
  { id: "boundary.glow@1", name: "boundary.glow", version: 1 },
  { id: "boundary.glow@2", name: "boundary.glow", version: 2 },
  { id: "kernel.rbf@1", name: "kernel.rbf", version: 1 },
  { id: "zetaid.glyph@1", name: "zetaid.glyph", version: 1 },
  { id: "zetaid.glyph@2", name: "zetaid.glyph", version: 2 },
];

const out: Record<string, string> = { _source: "generated-from-ir" };
for (const { id, name, version } of inputs) {
  out[id] = idOf(name, version);
}

const target = join(dirname(import.meta.dir), "ts-output.json");
writeFileSync(target, `${JSON.stringify(out, null, 2)}\n`);
console.log("wrote ts-output.json (generated-from-ir)");
